### ISSUE TITLE
class_embed 

### ISSUE THREAD

**Issue by Innary** [IDEA-opensource/DN-DETR#23]
`**self.class_embed = nn.Linear(hidden_dim, num_classes)**
  self.bbox_embed = MLP(hidden_dim, hidden_dim, 4, 3)
  self.num_feature_levels = num_feature_levels
  self.use_dab = use_dab
  self.num_patterns = num_patterns
  self.random_refpoints_xy = random_refpoints_xy
  # dn label enc
  **self.label_enc = nn.Embedding(num_classes + 1, hidden_dim - 1)** `
请问为什么这里class_embed使用的是91类，而label_enc使用的是92类？标准的DETR里class_embed里似乎是
`nn.Linear(hidden_dim, num_classes+1)`
Why class_embed  use class 91 here and label_enc use class 92?

**Comment by FengLi-ust** [comment]
Hey, because label_enc is also used to encode `empty object` (content query of the matching part), we add another category here. 
`tgt = label_enc(torch.tensor(num_classes).cuda()).repeat(num_queries * num_patterns, 1)`. 
You can refer to the `prepare_for_dn` function in`dn_components.py` for more details.

**Comment by Innary** [comment]
@FengLi-ust  Thanks for the answer~

我之前参考的是deformable DAB-DETR 的代码，并没有这个操作

1. dn_dab_deformable_detr/dn_components.py

`indicator0 = torch.zeros([num_queries * num_patterns, 1]).cuda()`
   
`tgt = torch.cat([tgt_weight, indicator0], dim=1) + label_enc.weight[0][0]*torch.tensor(0).cuda()`

标识符直接与tgt拼接

Identifiers are directly spliced with tgt

2. DN_DAB_DETR/dn_components.py 
`tgt = label_enc(torch.tensor(num_classes).cuda()).repeat(num_queries * num_patterns, 1)`
`tgt = torch.cat([tgt, indicator0], dim=1)`

所有tgt被encode成背景类

All tgt are encoded as background classes

请问为什么两者对于tgt有两种截然不同的操作

Why are there two different operations on tgt?

**Comment by FengLi-ust** [comment]
Because dab_detr uses all zero vectors as the initial content query embedding, therefore we directly embed them into the background category to distinguish from the dn part.
However, dab_deformable_detr uses learnable content query embedding, which contains information learned from the dataset. So we `cat` the indicator with the `tgt` to preserve the original content information.

The above two ways are just used to distinguish the dn part and original matching part, which do not have a significant impact on the performance. Thank you.

**Comment by Innary** [comment]
thank you for you anwser~ @FengLi-ust
